pkg/bpf/map_linux.go

```diff
@@ -941,34 +938,55 @@ func (m *Map) deleteMapEntry(key MapKey, ignoreMissing bool) (deleted bool, err
 		return false, err
 	}
 
-	_, errno := deleteElement(m.FD(), key.GetKeyPtr())
-	deleted = errno == 0
+	var duration *spanstat.SpanStat
+	if option.Config.MetricsConfig.BPFSyscallDurationEnabled {
+		duration = spanstat.Start()
+	}
 
-	// Error handling is skipped in the case ignoreMissing is set and the
-	// error is ENOENT. This removes false positives in the delete metrics
-	// and skips the deferred cleanup of non-existing entries. This situation
-	// occurs at least in the context of cleanup of NAT mappings from CT GC.
-	handleError := errno != unix.ENOENT || !ignoreMissing
+	err = m.m.Delete(key.GetKeyPtr())
 
-	if option.Config.MetricsConfig.BPFMapOps && handleError {
-		metrics.BPFMapOps.WithLabelValues(m.commonName(), metricOpDelete, metrics.Errno2Outcome(errno)).Inc()
+	if option.Config.MetricsConfig.BPFSyscallDurationEnabled {
+		metrics.BPFSyscallDuration.WithLabelValues(metricOpDelete, metrics.Error2Outcome(err)).Observe(duration.End(err == nil).Total().Seconds())
 	}
 
-	if errno != 0 && handleError {
-		err = fmt.Errorf("unable to delete element %s from map %s: %w", key, m.name, errno)
+	if errors.Is(err, ebpf.ErrKeyNotExist) && ignoreMissing {
+		// Error and metrics handling is skipped in case ignoreMissing is set and
+		// the map key did not exist. This removes false positives in the delete
+		// metrics and skips the deferred cleanup of nonexistent entries. This
+		// situation occurs at least in the context of cleanup of NAT mappings from
+		// CT GC.
+		return false, nil
 	}
-	return
+
+	if option.Config.MetricsConfig.BPFMapOps {
+		// err can be nil or any error other than ebpf.ErrKeyNotExist.
+		metrics.BPFMapOps.WithLabelValues(m.commonName(), metricOpDelete, metrics.Error2Outcome(err)).Inc()
+	}
+
+	if err != nil {
+		return false, fmt.Errorf("unable to delete element %s from map %s: %w", key, m.name, err)
+	}
+
+	return true, nil
 }
 
 // SilentDelete deletes the map entry corresponding to the given key.
```

Comment:
Drive by: this description doesn't tell me at all what makes the function so silent.

Reply:
Yeah, this needs to go. As you can see, the Map API is full of this sort of snowflake and unhelpful/incomplete doc. To answer your question: it skips incrementing the map's delete metric.

IMO this highlights a common problem with the design of the data model in the upper layer(s). The data (e.g. ct record) and the underlying data store (e.g. the ct map) are conflated. The bpf layer shouldn't lie about the amount of deletes it issues to the kernel, as it could mask a potential perf problem.